### PR TITLE
feat: task auto discovery wrapper for vcp cli

### PR DESCRIPTION
**Pull request by saugupta1**
## 1. Code Changes

- **clustering.py**, **embedding.py**, **integration.py**, **label_prediction.py**  
  Made small updates so the system can automatically find and register tasks. 

- **runner.py**  
  Added a new, functional-style API called **run_task** that lets you start tasks programmatically.

- **single_cell/** (cross_species.py, perturbation.py, __init__.py)  
  Made small updates so the system can automatically find and register tasks. 

- **task.py**, **types.py**, **utils.py**  
  Introduced a **TaskRegistry** and let each task register itself automatically.

## 2. Test Cases for runner.py

- Added tests to ensure **run_task** can launch every supported task.
- Verified baseline calculations and final results.
- Covered edge cases and errors when inputs are invalid.

## 3. Enhanced Documentation

- **tasks.md**  
  Updated instructions to show how to use the new **run_task** API.

- **add_new_task.md**  
  Revised steps for adding a new task, now using **run_task**.

### CODE REVIEW

File: src/czbenchmarks/cli/cli_run.py [hunk 1]
```diff
@@ -56,81 +49,196 @@ def run(ctx):
...
-    baseline_params = discover_baseline_parameters(task_def)
+for task_name in TASK_REGISTRY.list_tasks():
+    task_info = TASK_REGISTRY.get_task_info(task_name)
 
-    @click.command(name=task_name, help=task_def.description)
-    @add_options(task_params)
-    @add_options(baseline_params)
-    @common_task_options
+    @click.command(name=task_name, help=task_info.description)
     def task_callback(**kwargs):
```
Comment (steveherrin) [inline src/czbenchmarks/cli/cli_run.py:61]: This may also be a follow-up ticket, but `click` does a lot of the validation typecasting (including paths) for you if the command arguments have type annotations. Using `**kwargs` means you have to have the extra code to do it here. You have enough information from the task params to construct the actual signature for this, but it's a little tricky in terms of metaprogramming, so maybe for later.

File: src/czbenchmarks/cli/cli_run.py [hunk 2]
```diff
@@ -56,81 +49,196 @@ def run(ctx):
...
+                exc_info=True,
             )
-
-        write_results(all_metric_results, output_file)
-        click.echo(f"✅ Task '{invoked_task_name}' completed successfully.")
+            raise click.ClickException(f"An unexpected error occurred: {e}")
+
+    # --- Dynamically Add Options to the Command ---
+    def add_click_options(options):
+        def decorator(f):
```
Comment (steveherrin) [inline src/czbenchmarks/cli/cli_run.py:180]: should be
```
@functools.wraps(f)
def decorator(f)
```
That way the decorated function still retains its original name and doctring and stuff

File: src/czbenchmarks/tasks/task.py [hunk 3]
```diff
@@ -21,6 +29,294 @@ class TaskOutput(BaseModel):
     model_config = {"arbitrary_types_allowed": True}
 
 
+class TaskParameter(BaseModel):
+    """Schema for a single, discoverable parameter."""
+
+    type: str
```
Comment (steveherrin) [inline src/czbenchmarks/tasks/task.py:35]: Why not `type: type`?

File: src/czbenchmarks/tasks/task.py [hunk 4]
```diff
@@ -21,6 +29,294 @@ class TaskOutput(BaseModel):
...
+    task_params: Dict[str, TaskParameter]
+    baseline_params: Dict[str, TaskParameter]
+    metrics: List[str]
+
+
+class TaskRegistry:
+    """A registry that is populated automatically as Task subclasses are defined."""
+
+    def __init__(self):
+        self._registry: Dict[str, Type["Task"]] = {}
```
Comment (steveherrin) [inline src/czbenchmarks/tasks/task.py:55]: If you `from __future__ import annotations` you shouldn't need to quote all the forward declarations for `Task`

File: src/czbenchmarks/tasks/task.py [hunk 5]
```diff
@@ -21,6 +29,294 @@ class TaskOutput(BaseModel):
...
+        self._info: Dict[str, TaskInfo] = {}
+
+    def register_task(self, task_class: Type["Task"]):
+        """Registers a task class and introspects it to gather metadata."""
+        if inspect.isabstract(task_class) or not hasattr(task_class, "display_name"):
+            return
```
Comment (steveherrin) [inline src/czbenchmarks/tasks/task.py:61]: Should this be an error? Or maybe at least log an error? Otherwise someone might be confused as to why their new task isn't showing up when they forgot to set a display name

File: src/czbenchmarks/tasks/task.py [hunk 6]
```diff
@@ -21,6 +29,294 @@ class TaskOutput(BaseModel):
...
+                        if field_info.default is not PydanticUndefined
+                        else None,
+                        required=field_info.is_required(),
+                    )
+
+            # 2. Get Baseline Parameters from the compute_baseline method signature
+            baseline_params = {}
+            try:
+                sig = inspect.signature(task_class.compute_baseline)
+                for param in list(sig.parameters.values())[1:]:
```
Comment (steveherrin) [inline src/czbenchmarks/tasks/task.py:99]: Is the `[1:]` to skip `self`? I'd probably just have an `if param.name == "self": continue` so it's clearer to the reader. Or at least have a comment

File: src/czbenchmarks/tasks/task.py [hunk 7]
```diff
@@ -21,6 +29,294 @@ class TaskOutput(BaseModel):
...
+            return TaskInfo(
+                name=task_class.__name__,
+                display_name=getattr(task_class, "display_name", task_class.__name__),
+                description="Task introspection failed - please check task implementation",
+                task_params={},
+                baseline_params={},
+                metrics=[],
+            )
+
+    def _extract_type_info(self, annotation: Any, param_name: str) -> str:
```
Comment (steveherrin) [inline src/czbenchmarks/tasks/task.py:153]: As I continue to read, I still think storing the type information as a `type` would be easier. I don't think you'd need this at all anymore

File: src/czbenchmarks/tasks/task.py [hunk 8]
```diff
@@ -21,6 +29,294 @@ class TaskOutput(BaseModel):
...
+                        default=field_info.default
+                        if field_info.default is not PydanticUndefined
+                        else None,
+                        required=field_info.is_required(),
+                    )
+
+            # 2. Get Baseline Parameters from the compute_baseline method signature
+            baseline_params = {}
+            try:
+                sig = inspect.signature(task_class.compute_baseline)
```
Comment (steveherrin) [inline src/czbenchmarks/tasks/task.py:98]: Instead of `inspect.signature`, did you consider `typing.get_type_hints`?

Then for the type, you get an actual python type instead of string that you have to make sense of.

example:
```python
from __future__ import annotation
def f(x: str | None, y: int = 0) -> list[str]:
    return [x or "None"] * y

hints = typing.get_type_hints(f, include_extras=True)  # {'x': str | None, 'y': <class 'int'>, 'return': list[str]}
hints['return']("abc")  # ['a', 'b', 'c']
```

File: src/czbenchmarks/tasks/task.py [hunk 9]
```diff
@@ -45,6 +341,11 @@ def __init__(
         # FIXME should this be changed to requires_multiple_embeddings?
         self.requires_multiple_datasets = False
 
+    def __init_subclass__(cls, **kwargs):
+        """Automatically register task subclasses when they are defined."""
+        super().__init_subclass__(**kwargs)
+        TASK_REGISTRY.register_task(cls)
```
[the remark this answers is not in the document]
Comment (steveherrin) [reply src/czbenchmarks/tasks/task.py:347]: This was my suggestion. I like this approach because anytime anyone subclasses `Task` anywhere, even in their own code, it will still end up in the registry. This is the approach libraries like SQLAlchemy take.

> Consider that this could be implemented via Python introspection when the Registry is instantiated

How would you do this and allow users to create their own tasks without modifying cz-benchmarks?

File: src/czbenchmarks/cli/cli_run.py (hunk 1 above)
Comment (steveherrin) [reply src/czbenchmarks/cli/cli_run.py:61]: Oh, I see you're adding them with decorators below. Hmmm. Maybe that's fine and we just need to actually specify the types to click there.

File: src/czbenchmarks/tasks/task.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (atolopko-czi) [reply src/czbenchmarks/tasks/task.py:347]: Sounds good, I wasn't convinced it was worth changing. Thanks for the discussion.

File: src/czbenchmarks/cli/cli_run.py [hunk 10]
```diff
@@ -1,53 +1,46 @@
...
-    func = click.option(
-        "--dataset",
-        "-d",
-        "dataset_names",
-        required=True,
-        multiple=True,
-        help="Name of the dataset(s) to use.",
-    )
-    func = click.option(
-        "--model-embedding-path",
```
[the remark this answers is not in the document]
Comment (saugupta1) [reply src/czbenchmarks/cli/cli_run.py:31]: It was just used for testing. Are we exposing run in cz-benchmarks?

File: src/czbenchmarks/cli/cli.py [hunk 11]
```diff
@@ -28,8 +29,6 @@ def main(log_level: str):
 
 # Add subcommands to the main group
 main.add_command(list_cmd)
-
-# TODO: Disabled until further notice. The run command is not yet fully implemented.
```
[the remark this answers is not in the document]
Comment (saugupta1) [reply src/czbenchmarks/cli/cli.py:32]: Implemented. Do we expose it?

File: src/czbenchmarks/tasks/task.py [hunk 12]
```diff
@@ -21,6 +29,294 @@ class TaskOutput(BaseModel):
...
+
+    name: str
+    display_name: str
+    description: str
+    task_params: Dict[str, TaskParameter]
+    baseline_params: Dict[str, TaskParameter]
+    metrics: List[str]
+
+
+class TaskRegistry:
```
[the remark this answers is not in the document]
Comment (saugupta1) [reply src/czbenchmarks/tasks/task.py:53]: It would cause circular import problem.
